### ISSUE TITLE
Question about the optimized objective.

### ISSUE THREAD

**Issue by JacobYuan7** [kvablack/ddpo-pytorch#12]
Many thanks for conducting this excellent work. While I read this repo, I raised two questions on the optimized objective. 
1. Does the loss term in Eq. 4 align with the code implementation below? `advantages` corresponds to `r(x_0, c)` and `ratio` corresponds to the importance sampling term. But where is the gradient term `\nabla_\theta \log p_\theta\left(\mathbf{x}_{t-1} \mid \mathbf{c}, t, \mathbf{x}_t\right)`? Is it because taking gradient on ratio will implicitly generate the gradient term?
```python
# ppo logic
advantages = torch.clamp(
    sample["advantages"], -config.train.adv_clip_max, config.train.adv_clip_max
)
ratio = torch.exp(log_prob - sample["log_probs"][:, j])
unclipped_loss = -advantages * ratio
clipped_loss = -advantages * torch.clamp(
    ratio, 1.0 - config.train.clip_range, 1.0 + config.train.clip_range
)
loss = torch.mean(torch.maximum(unclipped_loss, clipped_loss))
```
2. `clip_range` is set to 1e-4 by default, which makes the `clipped_loss` close to `-advantages`. The following operation `torch.maximum(unclipped_loss, clipped_loss)`  will bound `loss` to be approximately greater than `-advantages`. What is the aim of setting such a small value of `clip_range`?

**Comment by kvablack** [comment]
1. Yes, that's correct -- the autodiff algorithm will calculate the gradient from the paper.
2. The purpose of the clip range is to impose a constraint on the policy -- intuitively, we don't want to follow any gradients w.r.t. a policy that is too far away from the policy that generated the data.

I would recommend [reading about PPO](https://spinningup.openai.com/en/latest/algorithms/ppo.html) for more information.
